### ISSUE TITLE
System.Net.Security 4.3.0 breaks AuthenticateAsClientAsync on Linux

### ISSUE THREAD

**Issue by ReimuHakurei** [dotnet/corefx#17218]
Hopefully this is the right issue tracker for this bug.

System.Net.Security 4.3.0 breaks, at the very least, authenticating with Apple's push notification servers in an unusual way -- but only on Linux. It still works fine when used on Windows.

Here is a simple C# program, which does the bare minimum I've found to trigger this bug. For obvious reasons, I cannot provide the certificate file, although I expect any other instances of certificate client authentication would likely also trigger the bug.

```c#
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Security.Authentication;
using System.Threading.Tasks;

namespace Test {
	class Program {
		static void Main() {
			Task.Run(async () => {
				int port = 2195;
				string hostname = "gateway.push.apple.com";

				X509Certificate2 clientCertificate = new X509Certificate2(System.IO.File.ReadAllBytes("certificate.pfx"));
				X509Certificate2Collection certificatesCollection = new X509Certificate2Collection(clientCertificate);

				TcpClient client = new TcpClient(AddressFamily.InterNetwork);
				await client.ConnectAsync(hostname, port);

				SslStream sslStream = new SslStream(client.GetStream());
	
				await sslStream.AuthenticateAsClientAsync(hostname, certificatesCollection, SslProtocols.Tls12, false);

				Console.WriteLine("Success.");
			}).Wait();
		}
		
	}
}
```

And here is a matching csproj file:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NETCore.Portable.Compatibility">
      <Version>*</Version>
    </PackageReference>
    <PackageReference Include="System.IO">
      <Version>*</Version>
    </PackageReference>
    <PackageReference Include="System.Net.Security">
      <Version>4.0.0</Version>
    </PackageReference>
    <PackageReference Include="System.Net.Sockets">
      <Version>*</Version>
    </PackageReference>
    <PackageReference Include="System.Security.Cryptography.X509Certificates">
      <Version>*</Version>
    </PackageReference>
  </ItemGroup>

</Project>
```

When System.Net.Security depends on 4.0.0, the program works as expected; it successfully authenticates, prints "Success!", then exits.

If you change that dependency to 4.3.0, it instead pauses, then exits the program with no console output during AuthenticateAsClientAsync, and never prints "Success!". No exceptions are thrown, it just seemingly cleanly exits, when it should not.

**Comment by ReimuHakurei** [comment]
I'll check top later, but it is of note that the system in question has 1GB of memory + 2GB of swap free, and no other library exhibits the issue.

Where can I find a nightly build of System.Net.Security? I don't see any on NuGet. Are binaries available, or do I need to compile it myself?

**Comment by bartonjs** [comment]
@weshaggard I'm not sure how one goes about getting nightlies for testing anymore, given the current state of packaging, et al.  Do you have any advice?

**Comment by weshaggard** [comment]
See https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md

**Comment by ianhays** [comment]
@ReimuHakurei Are you having issues with this on the latest packages?

**Comment by karelz** [comment]
We do not have actionable information and we believed it is already fixed in 2.0.
Feel free to reopen if you have evidence saying otherwise.
